app/common/validators/AddressFormResultsHandler.scala

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package common.validators
+
+import config.FrontendAppConfig
+import models.api.{Address, Country}
+import play.api.data.Form
+import play.api.i18n.Messages
+import play.api.mvc.Results.BadRequest
+import play.api.mvc.{AnyContent, Call, Request, Result}
+import uk.gov.hmrc.http.InternalServerException
+import views.html.CaptureInternationalAddress
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.Future
+
+@Singleton
+class AddressFormResultsHandler @Inject()(view: CaptureInternationalAddress)
+                                         (implicit appConfig: FrontendAppConfig) {
+  private val postcodeField = "postcode"
+  private val countryField = "country"
+  private val postcodeRequiredErrorKey = "internationalAddress.error.postcode.empty"
+  private val postcodeRequiredCountries = List("United Kingdom", "Isle of Man", "Guernsey", "Jersey")
+
+  def handle(
+              countries: Seq[Country],
+              headingMessageKey: String,
+              addressForm: Form[Address],
+              submitAction: Call,
+              saveContactDetails: Address => Future[Result],
+              name: Option[String] = None
+            )(implicit request: Request[AnyContent],
+              messages: Messages): Future[Result] = {
+
+    addressForm.fold(
+      formWithErrors => {
+        val maybeCountryName = formWithErrors(countryField).value
+        val finalForm = if (missingPostcode(maybeCountryName, formWithErrors(postcodeField).value)) {
+          formWithErrors.withError(postcodeField, postcodeRequiredErrorKey, getCountryCode(countries, maybeCountryName))
+        } else {
+          formWithErrors
+        }
+        Future.successful(BadRequest(view(finalForm, countries.flatMap(_.name), submitAction, headingMessageKey, name)))
+      },
+      internationalAddress => {
+        val maybeCountryName = internationalAddress.country.flatMap(_.name)
+        if (missingPostcode(maybeCountryName, internationalAddress.postcode)) {
+          Future.successful(BadRequest(view(
+            internationalAddressForm = addressForm.fill(internationalAddress).withError(
+              postcodeField, postcodeRequiredErrorKey, getCountryCode(countries, maybeCountryName)
+            ),
+            countries = countries.flatMap(_.name),
+            submitAction = submitAction,
+            headingKey = headingMessageKey,
+            name = name
+          )))
+        } else {
+          saveContactDetails(internationalAddress)
+        }
+      }
+    )
+  }
+
+  private def missingPostcode(maybeCountry: Option[String], maybePostcode: Option[String]) = {
+    maybeCountry.nonEmpty && postcodeRequiredCountries.contains(maybeCountry.get) && maybePostcode.isEmpty
+  }
+
+  private[validators] def getCountryCode(countries: Seq[Country], maybeCountryName: Option[String]): String = {
```

Comment:
Missed this earlier, sorry. We tend to try and avoid using nested pattern matching where possible to improve readability.

Ultimately, we're throwing exceptions if the data is missing (because we should have it in our countries JSON), so we don't need to be too verbose about the None cases.

This could be cleaned up by:

- Passing country name into this method as a string (so handling the optionality on line 60 - the form will return errors if the country is missing so we will always have it here)
- Using get or else where possible instead of matching. e.g.
`countries
        .find(_.name.contains(countryName))
        .map(_.code.getOrElse(throw new InternalServerException(s"Missing country code for '$countryName'")))`

-

Reply:
The formatting went a bit squiffy there, but the intent was to have .find() and .map() on separate lines so it's easier to follow

Reply:
Hey! yeah, sounds good. Will tidyup to remove the nested match blocks. 
However, we won't be able to pass in `countryName` as string to the country code lookup method. reason being, if the `formWithErrors` at line 50 is due to missing country then trying to do a `.get` on the form value will then throw a 500 when we want the else block to kick in for missing country to be reported as 400 as part of form validation. 

Have kept the signature to still take in optional country name so that above case is handled. But, tidied up to make it readable when we do country code handling.